server/conditions/boundary_collision_condition.go

```diff
@@ -9,9 +10,12 @@ import (
 type BoundaryCollisionCondition struct{}
 
 // Apply applies the condition to the game state
-func (c *BoundaryCollisionCondition) Apply(gameState *pogo.GameState, actionGroup []actions.Action) error {
+func (c *BoundaryCollisionCondition) Apply(gameState *pogo.GameState, actionGroup []actions.Action, step int) (int, error) {
+	if step > MaxConditionSteps {
+		return step, fmt.Errorf("failed to handle boudary collision, took more than %v steps", MaxConditionSteps)
```

Comment:
I added this concept of tracking the steps it takes to complete a condition.  This check right here will save us from infinite loops since this method is recursive.  I added this check to all 3 conditions we have right now.